src/ragger/firmware/versions.py

```diff
@@ -0,0 +1,80 @@
+"""
+   Copyright 2022 Ledger SAS
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+"""
+from enum import Enum
+from typing import Dict, Optional, Type
+
+from semver import VersionInfo
+
+
+class VersionManager(Enum):
+
+    @classmethod
+    def get_last(cls,
+                 major: Optional[int] = None,
+                 minor: Optional[int] = None,
+                 patch: Optional[int] = None,
+                 *args) -> VersionInfo:
+        # error cases: (None, 2) or (1, None, 4) or (None, 2, 2)
+        if (minor is not None and major is None):
+            raise ValueError("Minor version can not be pinned if major is not")
+        if (patch is not None and (major is None or minor is None)):
+            raise ValueError("Patch version can not be pinned if major AND minor are not")
+        if major is minor is None:
+            return max(v.value for v in cls)
+        if major is not None and minor is None:
+            return max(v.value for v in cls if v.value.major == major)
+        if major is not None and minor is not None and patch is None:
+            return max(v.value for v in cls if (v.value.major == major and v.value.minor == minor))
+        try:
+            return max(
+                v.value for v in cls
+                if (v.value.major == major and v.value.minor == minor and v.value.patch == patch))
+        except ValueError:
+            raise ValueError(f"No version {major}.{minor}.{patch} found")
```

Comment:
Why catching here and not above? Maybe you could let the default message?

Reply:
Without management, the raised error would be `ValueError: max() arg is an empty sequence` which is not very helpfull.
But you're right this should be crafted for this case, but also with major only, or major + minor.